File: Python/Deadlock Avoidance/bankers_algorithm.py
#!/usr/bin/env python

"""Banker's Algorithm

This script demonstrates Banker's algorithm for deadlock avoidance. User
inputs the current system state in the form of MAXIMUM and ALLOCATED matrices,
and TOTAL vector. 
The NEED matrix and AVAILABLE vector are calculated first. Then safety status
of the system and safe sequence are determined using the Safety algorithm. If
the system is deemed safe, there is an option to request additional resources,
which demonstrates the Resource-Request algorithm. If the system is safe after 
granting additional resources to a process, a new safe sequence is determined.

Python: 3.10.8

External packages: None

Author: Siddhant Tiwari (github.com/stiwari-ds)
"""

import sys
from typing import Tuple


def calculate_need(maximum: list, allocated: list) -> list:
    """Calculates the need matrix using:
        NEED = MAXIMUM - ALLOCATED
    with all the matrices having shape m x n, where:
        m = number of processes
        n = number of resources

    Args:
        maximum (list): 2-D matrix for maximum resources required by processes.
        allocated (list): 2-D matrix for current resources allocated to processes.

    Returns:
        list: 2-D matrix for resources required by processes to complete.
    """
    need_matrix = []
    num_process = len(maximum)
    num_resource = len(maximum[0])
    for p in range(num_process):
        process_need = []
        for r in range(num_resource):
            process_need.append(maximum[p][r] - allocated[p][r])
        need_matrix.append(process_need)
    return need_matrix


def calculate_available(total: list, allocated: list) -> list:
    """Calculates the availale vector by first finding sum of allocated
    resources for all processes and subtracting it from the total
    resources of each type.

    Args:
        total (list): 1-D vector of total resources of each type.
        allocated (list): 2-D matrix of current resources allocated to processes.

    Returns:
        list: 1-D vector of current available resources of each type.
    """
    available = []
    num_process = len(allocated)
    num_resource = len(total)
    for r in range(num_resource):
        allocated_sum = 0
        for p in range(num_process):
            allocated_sum += allocated[p][r]
        available.append(total[r] - allocated_sum)
    return available


def can_finish(process: int, need: list, available: list) -> bool:
    """Checks whether the process can be completed immediately by comparing
    its needed resources with the available vector.

    Args:
        process (int): process number
        need (list): 2-D matrix for resources required by processes to complete.
        available (list): 1-D vector for current available resources of each type.

    Returns:
        bool: whether the process can be completed immediately or not.
    """
    num_resource = len(available)
    for resource in range(num_resource):
        if need[process][resource] > available[resource]:
            return False
    return True


def check_safety(allocated: list, need: list, available: list) -> str:
    """Determines the safety state and safe sequence of the system based on
    currently allocated resources, resources needed by processes for completion,
    and the resources available for allocation. If no safe sequence can be determined,
    it means the system is in an unsafe state.

    Args:
        allocated (list): 2-D matrix of current resources allocated to processes.
        need (list): 2-D matrix for resources required by processes to complete.
        available (list): 1-D vector for current available resources of each type.

    Returns:
        str: Safe sequence for completion of all processes. None if system is unsafe.
    """
    safe_sequence = []
    num_process = len(allocated)
    num_resource = len(available)
    finished = [False] * num_process  # all processes unfinished initially

    # waiting for all processes to complete
    while sum(finished) < num_process:
        # keeps track of whether at least one process could be allocated
        allocation_flag = False
        for p in range(num_process):
            if not finished[p]:  # only checking unfinished processes
                if can_finish(p, need, available):
                    allocation_flag = True
                    safe_sequence.append(p)
                    finished[p] = True
                    # available resources updated after process completion
                    for r in range(num_resource):
                        available[r] += allocated[p][r]
        if allocation_flag:
            continue
        else:
            # if no process could be alloted resources in this cycle, system is deadlocked
            break

    if sum(finished) < num_process:
        return None
    else:  # safe sequence returned only if all processes are completed
        safe_sequence = [f"P{i}" for i in safe_sequence]
        return " -> ".join(safe_sequence)


def check_resource_request(
    allocated: list, need: list, available: list, process: int, request: list
) -> Tuple[bool, str]:
    num_resource = len(available)
    num_process = len(allocated)

    # Condition 1: requested resources should not exceed actual need
    for r in range(num_resource):
        if request[r] > need[process][r]:
            return (False, "Request is more than actual need. Denied.\n")

    # Condition 2: requested resources should not exceed available resources
    for r in range(num_resource):
        if request[r] > available[r]:
            return (False, "Request is more than available resources. Denied.\n")

    # if above two conditions are satisfied, assume the request is granted
    # and update the system state
    for r in range(num_resource):
        allocated[process][r] += request[r]
        need[process][r] -= request[r]
        available[r] -= request[r]

    # check safety with the assumption that request was granted
    safe_sequence = check_safety(allocated, need, available.copy())
    if safe_sequence:
        return (True, safe_sequence)
    else:
        return (False, "System unsafe after granting additional request. Denied.\n")


def bankers_algorithm():
    print("\n----Input system state----")
    try:
        num_resource = int(input("\nNumber of resources: ").strip())
        num_process = int(input("Number of processes: ").strip())
    except ValueError:
        sys.exit("Integer values expected.")

    maximum = []
    allocated = []
    total = []

    print("\nMAXIMUM:")
    for i in range(num_process):
        try:
            process_max = list(map(int, input(f"P{i}: ").strip().split()))
        except ValueError:
            sys.exit("Integer values expected.")
        maximum.append(process_max)

    print("\nALLOCATED:")
    for i in range(num_process):
        try:
            process_alloc = list(map(int, input(f"P{i}: ").strip().split()))
        except ValueError:
            sys.exit("Integer values expected.")
        allocated.append(process_alloc)

    try:
        total = list(map(int, input(f"\nTOTAL: ").strip().split()))
    except ValueError:
        sys.exit("Integer values expected.")

    print("\nCalculating need matrix and available vector...")
    need = calculate_need(maximum, allocated)
    available = calculate_available(total, allocated)
    print("\nNEED:")
    for p in range(num_process):
        print(f"P{p}: {need[p]}")

    print(f"\nAVAILABLE: {available}")

    print("\nChecking safe state...")
    safe_sequence = check_safety(allocated, need, available.copy())
    if safe_sequence:
        print(f"The system is in a safe state.\nSafe sequence: {safe_sequence}")
    else:
        print("The system is in an unsafe state.\n")

    # Resource-request algorithm demonstrated only for system in safe state
    if safe_sequence:
        choice = input("\nTry requesting additional resources? [y/n]: ").strip().lower()
        if choice == "y":
            try:
                process = int(input("Process number: ").strip())
                request = list(map(int, input("Request: ").strip().split()))
            except ValueError:
                sys.exit("Integer values expected.")
            still_safe, new_safe_sequence = check_resource_request(
                allocated, need, available, process, request
            )
            if still_safe:
                print(
                    f"System remains in a safe state after granting the additional request.\n"
                    f"New safe sequence: {new_safe_sequence}\n"
                )
            else:
                print(new_safe_sequence)


if __name__ == "__main__":
    bankers_algorithm()

File: Java/Deadlock Avoidance/Bankers_Algorithm.java
/*
 ---------------------------------------------- Banker's Algorithm -------------------------------------------------------

    A algorithm which is used for resource allocation and deadlock avoidance in operating system. It is based on the
    concept of bank by allocating money to the requested user the bank should tally itself with money left in bank,
    likewise it works.
    It has two kind :-
            1) Safety Type
            2) Resource Request Type.

 ------------------------------------------------- Safety Type -----------------------------------------------------------

    1) To start with we should be able to allocate resources to all the process by executing any of the process
       need not be in a sequence manner.

    2) For writing the safe sequence of the process we should know the allocation of resources, maximum allocation,
       available resources and lastly the need matrix.

                    need matrix = maximum matrix - allocation matrix.

       the need matrix should not be negative or else we cannot allocate the available resources to the process.

    3) If the available resources of the process is greater than or equal to the needed resources of the process we can
       allocate the resources to that process or else we cannot allocate on top of that the process we allocated the
       resources will be finished at a certain time so the allocated resources will be free for that process we have to sum it.

                    available[process][resources] += allocation[process][resources]

    4) Since the available matrix is modified now we have to check again we can allocate the resources to the process we can't
       allocate resources before. Atlast we cannot allocate resources till now then we can say that DEADLOCK occurs.

 --------------------------------------------- Resource Request Type -----------------------------------------------------

    1) We have to allocate the additional resources requested by the process if we allocate the resources the system should not
       enter into DEADLOCK state.

    2) The requested additional resources should be less than the available resources and the need matrix of that process.

                    temp[resources] < need[process_no][resources]
                    temp[resources] < available[process_no][resources]

    3) Secondly we have to update the available, need and the allocation matrix with the additional resources.

                    need[process_no][i]-=temp[i];
                    available[0][i]-=temp[i];
                    allocation[process_no][i]+=temp[i];

    4) Then we have to check that if we allocate these additional resources to that process then we will be in a safe state
       by using the safety algorithm above.
       
 ------------------------------------------------- Complexities ----------------------------------------------------------
 
    Time Complexity  :- BigO(n*n*m)
    Space Complexity :- BigO(n*m)

 -------------------------------------------------- Drawbacks ------------------------------------------------------------

    1) It requires a fixed number of processes, and no additional processes can be started in the system
       while executing the process.

    2) The algorithm does no longer allows the processes to exchange its maximum needs while processing its tasks.

    3) Each process has to know their maximum resource requirement before starting the system.

*/

// Importing the Scanner class to read the input from user.
import java.util.Scanner;
public class Bankers_Algorithm
{
    // Declaring global variables, so we can use anywhere in the program.
    static Scanner sc;
    static int process;
    static int resources;
    static int allocation[][];
    static int maximum[][];
    static int available[][];
    static int temp_available[][];
    static int need[][];
    static int total[][];
    static int safesequence[];
    static boolean visited[];

    // The main method.
    public static void main(String[] args)
    {
        // Initializing he scanner class.
        sc = new Scanner(System.in);

        // Reading the number of process and number of resources from user.
        System.out.print("Enter number of process = ");
        process = sc.nextInt();
        System.out.print("Enter number of resources = ");
        resources = sc.nextInt();

        // Initializing all the declared variables.
        // The allocation matrix
        allocation = new int[process][resources];

        // The maximum matrix
        maximum = new int[process][resources];

        // The available matrix
        available = new int[process][resources];

        // Storing the available matrix in another matrix because the available matrix changes everytime, when the process allocated to bring
        // it to original available matrix this temp_available matrix will be used.
        temp_available = new int[process][resources];

        // The total resources going to be used by this process
        total = new int[1][resources];

        // The need matrix
        need = new int[process][resources];

        // The boolean visited matrix says about the process is allocated with the resources or not.
        visited = new boolean[process];

        // Lastly the answer matrix or the safe sequence matrix.
        safesequence = new int[process];

        // Reading the matrix inputs using this method.
        reading_input(process,resources);
    }

    // Reads the matrix inputs from the user
    private static void reading_input(int process, int resources)
    {
        // Reading the allocation matrix
        System.out.println("Allocation matrix,");
        for(int i=0;i<process;i++)
        {
            for(int j=0;j<resources;j++)
            {
                allocation[i][j] = sc.nextInt();
            }
        }

        // Reading the maximum matrix
        System.out.println("Maximum matrix,");
        for(int i=0;i<process;i++)
        {
            for(int j=0;j<resources;j++)
            {
                maximum[i][j] = sc.nextInt();
            }
        }

        // Calculating the need matrix
        // need matrix = maximum matrix - allocation matrix.
        for(int i=0;i<process;i++)
        {
            for(int j =0;j<resources;j++)
            {
                need[i][j] = maximum[i][j] - allocation[i][j];
                // The need matrix cannot be negative, if it is then we cannot allocate the resources.
                if(need[i][j] < 0)
                {
                    System.out.println("The allocation of resources are higher than the maximum resources");
                    System.out.println("It can't calculate the need matrix");
                    System.exit(0);
                }
            }
        }

        // Giving the choice to the user about if they know about any of these values we can allocate the resources.
        System.out.println("If you know available resources value then please enter 1 or if you know the total resources value" +
                " then enter 2");

        // Reading the choice of the user.
        int response = sc.nextInt();

        // Based on the choice entered the available resources will be calculated.
        switch (response)
        {
            // Directly reading the available resources from the user if user knows about it.
            case 1:
                System.out.println("Please enter the available resource value ");
                for(int i=0;i<resources;i++)
                {
                    available[0][i] = sc.nextInt();
                }
                break;
            // The user knows about total resources only so we are calculating the available resources from total resources.
            case 2:
                System.out.println("Please enter the total resource value ");
                for(int i=0;i<resources;i++)
                {
                    total[0][i] = sc.nextInt();
                }

                // Calculating the value of available matrix from the total and allocation.
                // Available matrix = total resources matrix - total allocated resources matrix.
                // Temporary array to store the total allocation matrix.
                int temp[][] = new int[1][resources];
                for(int j =0; j< resources;j++)
                {
                    int sum =0;
                    for(int i =0; i<process;i++)
                    {
                        sum+=allocation[i][j];
                    }
                    temp[0][j] = sum;
                }

                // Available matrix = total resources matrix - total allocated resources matrix.
                for(int i=0;i<resources;i++)
                {
                    available[0][i] = total[0][i] - temp[0][i];
                }
                break;
        }

        // Printing the readed value to the user before going to banker's algorithm
        printingvalues();
    }

    // Printing the readed values.
    private static void printingvalues()
    {
        // Printing the allocation matrix.
        System.out.println("Allocation Matrix :-");
        for(int i=0;i<process;i++)
        {
            for(int j=0;j<resources;j++)
            {
                System.out.print(allocation[i][j] + " ");
            }
            System.out.println();
        }

        // Printing the maximum matrix.
        System.out.println("Maximum Matrix :-");
        for(int i=0;i<process;i++)
        {
            for(int j=0;j<resources;j++)
            {
                System.out.print(maximum[i][j] + " ");
            }
            System.out.println();
        }

        // Printing the available resources.
        System.out.println("The available resources,");
        for(int i=0;i<resources;i++)
        {
            // The available resources cannot be less than 0, if it was then we can't allocate the resources to any tasks.
            if(available[0][i] <0)
            {
                System.out.println("The process is not having any available resource to allocate these tasks");
                System.exit(0);
            }
            System.out.print(available[0][i]+ " ");
            // Storing the available resources to temp_available resources matrix.
            temp_available[0][i] = available[0][i];
        }

        // Printing the need matrix.
        System.out.println("\nThe need matrix :-");
        for(int i=0;i<process;i++)
        {
            for(int j=0;j<resources;j++)
            {
                System.out.print(need[i][j] + " ");
            }
            System.out.println();
        }

        // Calling the banker's algo method to compute the banker's algorithm.
        bankersAlgo();
    }

    // Method which performs the banker's algorithm of SAFETY kind.
    private static void bankersAlgo()
    {
        // Counter variable to track, how many times the loop runs which should be less than the total process.
        int counter = 0;
        // The variable which says about the deadlock or the safe sequence.
        int l =0;
        // use loop to iterate, until all the processes are not finished.
        while (counter < process)
        {
            // find in-finished process which needs can be satisfied with the current available resource.
            boolean foundSafeSystem = false;
            // The banker's algo of first kind.
            for (int m = 0; m < process; m++)
            {
                // when process is not finished
                if (visited[m] == false)
                {
                    int j;
                    // Loop to check whether the need of each process for all the resources is less than the available resource.
                    for (j = 0; j < resources; j++)
                    {
                        //check need of current resource for current process with available.
                        if (need[m][j] > available[0][j])
                        {
                            break;
                        }

                    }
                    // j and total resources will be equal when all the needs of current process are satisfied
                    if (j == resources)
                    {
                        for (int k = 0; k < resources; k++)
                        {
                            available[0][k] += allocation[m][k];
                        }
                        // Add current process in the safesequence array
                        safesequence[counter++] = m;
                        // Make this process finished
                        visited[m] = true;
                        foundSafeSystem = true;
                    }
                }
            }
            // The condition of DeadLock.
            if (foundSafeSystem == false)
            {
                l = 1;
                System.out.println("The system is not in the safe state");
                System.out.println("----- DEADLOCK OCCURS -----");
                break;
            }
        }
        // Printing the safe sequence of the tasks.
        if(l!=1)
        {
            System.out.println("The safe sequence of the system is : ");
            for (int i = 0; i < process; i++)
            {
                if(i==process-1)
                {
                    System.out.println("P" + safesequence[i]);
                }
                else
                {
                   System.out.print("P" + safesequence[i] + " --> ");
                }
            }
        }
        // The second variety of banker's algo.
        // The resource request kind starts here.
        // Giving choice to user whether they want any additional resources to a process or not.
        System.out.println("\nDo you want any additional resources to be allocated to a process");
        System.out.println("Yes : 1, No : 2");

        // Reading the response from the user
        int response = sc.nextInt();

        // Performing based on the response by the user.
        switch (response)
        {
            // The user wants to allocate resources to a process.
            case 1:
                // Since the available matrix is changed by allocating the resources to all tasks.
                // We have to change it to original value.
                for(int i =0;i<resources;i++)
                {
                    available[0][i] = temp_available[0][i];
                }
                // Since the boolean array is changed to true we have to set it to false. To say whether the deadlock occurs or not.
                for(int i=0;i<process;i++)
                {
                    visited[i] = false;
                }
                // Calling the resource request method.
                resource_request();
                break;
            // The user doesn't want to allocate any resources to a process.
            case 2:
                // The program will be terminated.
                System.out.println("The program is terminated");
                System.exit(0);
                break;
        }
    }

    // The method which performs the resource request.
    private static void resource_request()
    {
        // Reading the process number which user wants to allocate additional resources.
        System.out.println("Enter the process number which you want additional resources");
        int process_no = sc.nextInt();

        // Temporary array to store the additional resources of that process.
        int temp[] = new int[resources];

        // Reading the additional resources values from the user.
        System.out.println("Enter the additional resources of that process");
        for(int i=0;i<resources;i++)
        {
            temp[i] = sc.nextInt();
            // The additional resources requested should be less than need matrix.
            if(temp[i] > need[process_no][i])
            {
                System.out.println("The additional resources you requested is exceeding the needed resources so I can't allocate" +
                        " or else system will enter into a UNSAFE STATE");
                System.exit(0);
            }
            // The additional resources requested should be less than available matrix.
            else if(temp[i] > available[0][i])
            {
                System.out.println("The additional resources you requested is exceeding the available resources so I can't allocate" +
                        " or else system will enter into a UNSAFE STATE");
                System.exit(0);
            }
        }

        // Updating the need, available, allocation matrix values to allocate the additional resources.
        for(int i=0;i<resources;i++)
        {
            need[process_no][i]-=temp[i];
            available[0][i]-=temp[i];
            allocation[process_no][i]+=temp[i];
        }

        // Calling the banker's algo of first kind to check that it will not cause any deadlock after allocating the additional resources.
        System.out.println("After allocating the resources you requested,");
        bankersAlgo();
    }
}
